### ISSUE TITLE
-Wvla should not fire for function array parameters

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#57098]
https://www.open-std.org/jtc1/sc22/wg14/www/docs/n2992.pdf was adopted into C2x and makes it more clear that VLA syntax in a function parameter list is not actually a VLA because the parameters are adjusted to be pointers, and so the parameter type is a pointer type, not a VLA type.

This means we should not issue a `-Wvla` warning for code like `void func(int n, int m[n]);` as in: https://godbolt.org/z/q9x6GxYM1.

Note, while this is a change for C2x, it's clarifying wording that has existed since VLAs were introduced, so I do not think this adjustment should only happen for C2x.

**Comment by uecker** [comment]
There is a similar bug for GCC:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98217

**Comment by tstanisl** [comment]
Actually it should not even fire for declaration of a pointer to VLA.
```
int main() {
    int a = 5;
    int (*p)[a];
    (void)p;
}
```

prints spurious warning:
```
<source>:3:14: warning: variable length array used [-Wvla]
    int (*p)[a];
```

**Comment by AaronBallman** [comment]
> Actually it should not even fire for declaration of a pointer to VLA.
> 
> ```
> int main() {
>     int a = 5;
>     int (*p)[a];
>     (void)p;
> }
> ```
> 
> prints spurious warning:
> 
> ```
> <source>:3:14: warning: variable length array used [-Wvla]
>     int (*p)[a];
> ```

By my reading of the standard, that is a VLA. The type-specifier is `int`, the first `declarator` following the type specifier is an `array-declarator` whose `direct-declarator` is `( * identifier )` followed by `[ assignment-expression ]`. The `assignment-expression` is not an integer constant expression, which makes the array declarator a VLA. So the pointer itself is a variably modified type (C2x 6.7.6p3) because the full declarator contains a VLA (C2x 6.7.6.2p4).

Do you have another reading of the standard which makes that not a VLA?

**Comment by tstanisl** [comment]
Yes, but this logic would could be applied to `void func(int n, int m[n]);` as well. In the C standard the term VLA refer not to the object but rather to the **type** of this object. No VLA object is created in `int (*p)[a];`. For example:
```
int n = 5;
int A[5] = {1,2,3,4,5};
int (*p)[5] = &A;
int x = (*p)[2];
```
is a perfectly valid C99 code with **no** runtime-sized object created on stack. So there should be no warning here.

The parameter `int m[n]` declares VLA type of `int[n]`. This type  is transformed to `int *`. But the VLA type is formed anyway. If the size expression had side-effect (like `int m[n++]`) then the expression has to evaluated, so it is an error to pretend that there was no VLA there. If the parameter would be a 2d array like `void foo(int n, int a[n][n])` then it would be adjusted to `void foo(int n, int (*a)[n]);`. Should it still raise a warning or not? IMO, it should not.

The upcoming C23 standard will make VLAs and all derived types *mandatory* again while keeping creation VLAs with *automatic storage* as an optional feature. IMO, `-Wvla` option should reflect this logic. The other flag like `-Wvmt`  should be added (from Variably Modified Types)

**Comment by inclyc** [comment]
> The upcoming C23 standard will make VLAs and all derived types _mandatory_ again while keeping creation VLAs with _automatic storage_ as an optional feature. IMO, `-Wvla` option should reflect this logic. The other flag like `-Wvmt` should be added (from Variably Modified Types)

FWIW there are some discussions about this, see https://reviews.llvm.org/D132952, and the planned changes are not implemented yet (sorry for this)

**Comment by tstanisl** [comment]
Ok.. So basically it means that this issue report should be closed because the "-Wvla" address the declaration of any Variably Modified Type (including VLA type). That includes types of function parameters even though the type is adjusted to a pointer (possibly VMT).

Thus add lines below should generate a warning with `-Wvla`:

```
int n = 5;

int A[n]; // automatic VLA
int (*pA)[n]; // pointer to VLA
typedef int T[n]; // alias for VLA type
void foo(int A[n]); // forms VLA type int[n] that is adjusted to int*
void foo(int (*pA)[n]);
(int(*)[n]) NULL; // cast to pointer to VLA
```

The new flag `-Wvla-stack-allocation` will raise the warning only for `int A[n];`.

**Comment by mizvekov** [comment]
I don't understand why the conclusion here is that we should emit any warnings at all for types where variably modified appears within sugar only.

Can't we just fix all those places so that we check for VM on the canonical type only? And thus ignore VM appearing on the original type of semantic adjustments, on unused template arguments, and such.

**Comment by tstanisl** [comment]
@mizvekov, the problem is that VLA type is actually formed here. And the semantics of `-Wvla` indicate that it should detect all creations of VLA types. Take a look on code:
```
int n = 0;
void foo(int a[n++]) {
  ...
}
```
the definition of `foo` is *not* equivalent to:
```
void foo(int *a) {
  ...
}
```
because global variable `n` is incremented each time `foo` is called.
The function behaves like:
```
void foo(int *a) {
  typedef int __unnamed_type[n++]; 
  ...
}
```
The VLA type is formed so the warning should be raised.

**Comment by uecker** [comment]
Last time we discussed this it seems there was agreement that both GCC and Clang should change -Wvla to only warn about proper VLA objects on the stack as the most useful warning. Then there should be another warning about all uses of VLA types and VM-types (which are constructed from VLA types) for people who want to screen for compatibility with C89 or C++. I am not sure whether a new command-line flag is needed for the later uses case, as there are existing warning flags for compatibility.

**Comment by mizvekov** [comment]
A general check about any VLA usage at all should probably happen early when we finished parsing and are building VariableArray types, probably in `Sema::BuildArrayType`.

Otherwise sprinkling around checks for VM types seems messy, so I don't think considering VM for this problem is an efficient approach.

**Comment by AaronBallman** [comment]
> Last time we discussed this it seems there was agreement that both GCC and Clang should change -Wvla to only warn about proper VLA objects on the stack as the most useful warning. Then there should be another warning about all uses of VLA types and VM-types (which are constructed from VLA types) for people who want to screen for compatibility with C89 or C++. I am not sure whether a new command-line flag is needed for the later uses case, as there are existing warning flags for compatibility.

We discussed further in https://reviews.llvm.org/D132952 and the consensus on that was:

> Add two new warning groups (we can bikeshed the names later): `-Wvla-portability` and `-Wvla-stack-allocation` grouped under the existing `-Wvla`. The behavior of `-Wvla` won't change, but users can opt into the kind of specific diagnostic they want. `void func(int n, int array[n]);` would trigger `-Wvla-portability` but not `-Wvla-stack-allocation`, and `void func(int n) { int array[n]; }` would trigger both diagnostics (though we could debate having it skip the `-Wvla-portability` warning whenever giving any other "you're using a VLA" kind of warning).

Based on that, I see `int n = foo(); int (*pA)[n];` issuing a `-Wvla-portability` diagnostic but not a `-Wvla-stack-allocation` diagnostic.

**Comment by Origami404** [comment]
I am interested in it and may send a patch after one day or two.

**Comment by inclyc** [comment]
Patch https://reviews.llvm.org/D137343. Thanks @Origami404 !

**Comment by uecker** [comment]
I am not too deeply impressed by the fact that clang decided to diverge from GCC.

**Comment by AaronBallman** [comment]
> I am not too deeply impressed by the fact that clang decided to diverge from GCC.

Has GCC finalized their behavior here? https://gcc.gnu.org/bugzilla/show_bug.cgi?id=98217 looks to still be open.

**Comment by uecker** [comment]
Sorry, I should have said: ... diverges from what was agreed with GCC (and actually proposed from the clang side).  I was about to implement that it GCC, but now I am lost. I also think the real issue here is the default behavior of -Wvla which should improve, and the lesser problem is to give the user more options.

**Comment by AaronBallman** [comment]
> Sorry, I should have said: ... diverges from what was agreed with GCC (and actually proposed from the clang side).

Can you show me where this discussion was? The closest I can find is https://bugs.llvm.org/show_bug.cgi?id=48460 (linked to from the GCC bug) and the result of that sounds quite similar to what we concluded we have consensus on here about stack vs non-stack usage being an important split. (Sorry if I'm being dense and forgetting something I should be remembering -- I've had a lot of conversions about VLAs in Clang.)

> I was about to implement that it GCC, but now I am lost. I also think the real issue here is the default behavior of -Wvla which should improve, and the lesser problem is to give the user more options.

I think they're both related in some ways. It's reasonable to improve the behavior of -Wvla if we can agree on what "improved" means, but I'm wary of doing so in a way that users can't get back to equivalent behavior. We have some users who really like the behavior of -Wvla as it is, and we have other users who really don't like -Wvla as it is, so we're not going to make everyone happy no matter what we do. But taking away warnings users find useful doesn't seem like the correct approach, which is why I think we do need to consider how much granularity we think is reasonable.

(Btw, I'm *really* glad we're having these discussions before you started your implementation work in GCC and before we committed anything in Clang -- that's really lucky timing, so thank you *very* much for chiming in on this bug report!)

**Comment by uecker** [comment]
I was referring to the GCC bug tracker where Richard suggested to change -Wvla to only warn about stack usage. I did not work on it because for GCC because Martin S said he would be working on, but it seems he stopped. But I see now that it is also not so clear that changing -Wvla was what everyone had in mind. Adding options for the various use cases as the proposed, here is reasonable, although I think the compatibility warnings should then be part of -Wc90-c99-compat /  -Wc++-compat which GCC has (not sure about clang). For me it would then make the most sense to also change the default of -Wvla  to only warn about the stack case (as Richard suggested)  what seems to be the main motivation for using this flag. In any case, please add a comment to the GCC bug tracker if you make any change.

**Comment by AaronBallman** [comment]
@uecker -- FWIW, there's a patch up at https://reviews.llvm.org/D137343 for this (feel free to express your opinions on the patch as well, as that's where the decision will be made about what behavior to go with).

**Comment by kees** [comment]
Hi! It looks like this was never resolved even though there was a patch for implementing `-Wvla-stack-allocation`. Currently the Linux kernel uses `-Wvla` to make sure stack VLAs are never introduced, but we'd like to start exploring using function array parameters. GCC has `-Wvla-larger-than=0` which provides this coverage. What should Clang do to gain a way to distinguish between the two cases?

e.g.:
```
$ gcc -O2 -Wall -std=c99 -Wvla -c test.c -o test.o
test.c:3:1: warning: ISO C90 forbids variable length array ‘input’ [-Wvla]
    3 | void func(int n, int input[n])
      | ^~~~
test.c: In function ‘func’:
test.c:5:9: warning: ISO C90 forbids variable length array ‘array’ [-Wvla]
    5 |         int array[n];
      |         ^~~
$ gcc -O2 -Wall -std=c99 -Wvla-larger-than=0 -c test.c -o test.o
test.c: In function ‘func’:
test.c:5:13: warning: argument to variable-length array may be too large [-Wvla-larger-than=]
    5 |         int array[n];
      |             ^~~~~
```
